### ISSUE TITLE
ResNet models in tf.keras.applications contain a bias term which should not be there.

### ISSUE THREAD

**Issue by ujjwal-ai** [tensorflow/tensorflow#37365]
Pretrained ResNet models available as part of `tf.keras.applications` include a `bias` weight with all the convolutional layers which is weird. What is even weirder is that the pretrained weights contain `all zeros` for the bias weights, which is definitely a problem. 

ResNet models do not use a bias term because of the use of `batch normalization`. Even the TensorFlow models repository, the [ResNet construction code](https://github.com/tensorflow/models/blob/master/official/vision/image_classification/resnet_model.py) does not contain a bias term in convolutions.

The following code shows the weights in the resnet50 model in `tf.keras.applications` as can be seen the bias terms are all zeros.
```

import tensorflow as tf

model = tf.keras.applications.resnet50.ResNet50(include_top=False, weights='imagenet')
print(model.trainable_variables)
```

A small portion of the output shows the bias terms as all zeros.

```
array([[[[ 1.76406968e-02,  2.18379945e-02,  6.38491847e-03, ...,
          -1.56918354e-02,  1.33828130e-02, -7.58931879e-03],
         [ 6.57748384e-03, -1.13832625e-02, -1.44122150e-02, ...,
           1.07535999e-02,  1.99317057e-02, -5.90330362e-03],
         [ 1.96981058e-02,  6.84878789e-03, -1.30715151e-03, ...,
          -8.99719913e-03,  1.00973761e-02, -1.09837623e-02],
         ...,
         [ 4.02560830e-03, -2.51277094e-03, -1.91410668e-02, ...,
           1.84022412e-02, -1.05592925e-02,  3.84159223e-03],
         [-1.21582337e-02, -2.44973949e-03, -8.21000524e-03, ...,
          -3.52650182e-03,  9.62345582e-03, -1.55217517e-02],
         [-1.57500952e-02, -5.96316298e-03,  4.53999359e-03, ...,
           4.88574570e-03,  4.60040662e-03,  8.99072620e-05]]]],
      dtype=float32)>, <tf.Variable 'conv5_block3_3_conv/bias:0' shape=(2048,) dtype=float32, numpy=array([0., 0., 0., ..., 0., 0., 0.], dtype=float32)>, <tf.Variable 'conv5_block3_3_bn/gamma:0' shape=(2048,) dtype=float32, numpy=
array([1.3451786, 1.3965728, 1.4453218, ..., 1.2092956, 1.5969722,
       1.4321095], dtype=float32)>, <tf.Variable 'conv5_block3_3_bn/beta:0' shape=(2048,) dtype=float32, numpy=
array([-1.4512578, -1.6519743, -1.6319023, ..., -1.6061822, -1.7218091,
       -1.9375533], dtype=float32)>]

```

This is definitely an issue which needs to be cleared up as a lot of people are depending upon `tf.keras.applications`.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.1.0, 2.2.0-dev20200305 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/fe1cc7f0f62c04d52390653aa1d598df/untitled707.ipynb). Thanks!

**Comment by ujjwal-ai** [comment]
This is a rather serious issue requiring an urgent fix because too many people rely upon `tf.keras.applications` for pretrained models. As of yet there are no V2 versions of slim models available and hence `tf.keras.applications` is the only resort. 

Hence, unless this issue is resolved, usage of TF V2 for code porting and new code writing is going to be a big burden.

**Comment by jaketae** [comment]
I think you might be right, as the Kaiming He himself said [here](https://github.com/KaimingHe/deep-residual-networks/issues/10) that batch normalization should do away with the need for bias in the convolutional layers. If I understand it correctly, fixing this issue would just be a simple matter of setting `use_bias = False` when exporting the ResNet models. I can open a PR and start working soon. Can someone confirm that this is the right direction?

**Comment by ujjwal-ai** [comment]
@jaketae It seems a fair point to start. But the inconsistency does not end here. A major difference between V1 and V2 models in ResNet is the use of BN in V2 while V1 does not use BN. If the current models are observed, it is clear that they both have BN layers.

**Comment by jaketae** [comment]
@ujjwal-ai I'm not sure on your second point on BN. The [original paper] (https://arxiv.org/pdf/1512.03385.pdf) on ResNet states that BN is used "right after each convolution and before activation."

**Comment by ujjwal-ai** [comment]
Then what is the exact difference between V1 and V2? Maybe I am wrong here. So any help would be appreciated

**Comment by tanzhenyu** [comment]
@ujjwal-ai we should probably remove and upload a new checkpoint. But I'm not sure how this is blocking you from using keras applications model?

**Comment by ujjwal-ai** [comment]
This is not stopping me. I can use it by restricting the variables to be updated. The moot point is that for a new person already unaware of this issue, it can be dangerously confusing if he/she is not careful.

**Comment by jaketae** [comment]
@tanzhenyu Do you think the default export setting for ResNet should be changed so that `use_bias` is set to `False`?

**Comment by ujjwal-ai** [comment]
@jaketae sorry for jumping here but yes I think that 'use_bias=False' is the right thing to do.

**Comment by tiagofrepereira2012** [comment]
Hi,

Is there any fix on the way to this issue?
I can't do `tf.keras.applications.ResNet50(use_bias=False)`, since `use_bias` is a hardcoded set to `True` here https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/applications/resnet.py#L471

Of course, I can hack the way through it but would be nice to have a fix.

Thanks

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/6048195d58584b781d0dfd72c125a6da/35650.ipynb). Thanks!

**Comment by seermer** [comment]
Any updates on this? This problem still exists in the newest version. Extra bias terms can lead to extra unnecessary calculations and params. In fact, even for ResNet50, extra bias terms lead to more than 10k extra parameters!

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
